internal/controllers/machineset/machineset_controller.go

```diff
@@ -359,6 +329,79 @@ func (r *Reconciler) reconcile(ctx context.Context, cluster *clusterv1.Cluster,
 	return result, nil
 }
 
+func (r *Reconciler) reconcileDelete(ctx context.Context, machineSet *clusterv1.MachineSet) error {
+	log := ctrl.LoggerFrom(ctx)
+	machineList, err := r.getAndAdoptMachinesForMachineSet(ctx, machineSet)
+	if err != nil {
+		return err
+	}
+
+	// If all the descendant machines are deleted, then remove the machineset's finalizer.
+	if len(machineList) == 0 {
+		controllerutil.RemoveFinalizer(machineSet, clusterv1.MachineSetFinalizer)
+		return nil
+	}
+
+	log.Info("MachineSet still has descendant Machines - deleting them first", "count", len(machineList), "descendants", descendantMachines(machineList))
+
+	// else delete owned machines.
+	for _, machine := range machineList {
+		if machine.DeletionTimestamp.IsZero() {
+			log.Info("Deleting Machine", "Machine", klog.KObj(machine))
+			if err := r.Client.Delete(ctx, machine); err != nil {
+				return errors.Wrapf(err, "failed to delete Machine %s", klog.KObj(machine))
+			}
+		}
+	}
+
+	return nil
+}
+
+func (r *Reconciler) getAndAdoptMachinesForMachineSet(ctx context.Context, machineSet *clusterv1.MachineSet) ([]*clusterv1.Machine, error) {
```

Comment:
I renamed this from `getAndAdoptMachinesForMachineSet` to `getMachinesForMachineSet`. 

Question: 
* Should we keep `getMachinesForMachineSet` instead?
* Or should we adapt and also rename the equivalent function in the machinedeployment controller?


Some more context: https://github.com/kubernetes-sigs/cluster-api/pull/10791#discussion_r1651945666

Reply:
TODO: Also rename in machinedeployment controller ([source](https://github.com/kubernetes-sigs/cluster-api/pull/10791#discussion_r1755122021))

Reply:
+1 for me